### DISCUSSION TITLE
Current implementation of Order.subtotal

### DISCUSSION THREAD

**Discussion by carlosa54** [saleor/saleor#12395] [Q&A] [answered]
Hi Saleor team 👋 

While working with [Bug: Incorrect tax calculation in Order and Checkout total calculations](https://github.com/saleor/saleor/issues/12348) I noticed a strange behavior in the way that the Order subtotal is currently implemented and some bugs it that causes.

- The logic in [resolve_subtotal](https://github.com/saleor/saleor/blob/d6a9fbe84e97a36e64841a27a6eb5e05a53e8e4f/saleor/graphql/order/types.py#L1350-L1358) for `order.subtotal` is just recalculating the `order.total` https://github.com/saleor/saleor/blob/d6a9fbe84e97a36e64841a27a6eb5e05a53e8e4f/saleor/graphql/order/types.py#L1350-L1358

- There's a display bug in saleor-dashboard for the subtotal amount, given the current current logic it should be using `.net` instead of `.gross` to show the subtotal when there are taxes. (I can create a separate issue for this in saleor-dashboard)

    - [OrderPayment](https://github.com/saleor/saleor-dashboard/blob/5dabfc315fda1c871dab03da9b514a144f6c5b62/src/orders/components/OrderPayment/OrderPayment.tsx#L153)
    - [OrderDraftDetailsSummary](https://github.com/saleor/saleor-dashboard/blob/5dabfc315fda1c871dab03da9b514a144f6c5b62/src/orders/components/OrderDraftDetailsSummary/OrderDraftDetailsSummary.tsx#L205)
    - [OrderSummaryCard](https://github.com/saleor/saleor-dashboard/blob/5dabfc315fda1c871dab03da9b514a144f6c5b62/src/orders/components/OrderSummaryCard/OrderSummaryCard.tsx#L56)

I saw that the current implementation uses `TaxedMoney` which includes `tax, net and gross` here: https://github.com/saleor/saleor/blob/d6a9fbe84e97a36e64841a27a6eb5e05a53e8e4f/saleor/graphql/order/types.py#L1067-L1071
By definition the order subtotal should always be equal to the `order.total.net.amount` and given this my first thought is that the field definition should be of type `Money` instead of `TaxedMoney`. 

I wanted to confirm if the current behavior is expected, thanks in advance 😃

**Comment by korycins** [comment]
Hey @carlosa54 
By definition, the subtotal in ecommerce refers to the sum of all line prices multiplied by their respective quantities. The total, on the other hand, is the subtotal plus the shipping price.

I have never encountered a situation where the subtotal means the net of the total, but it is possible that naming conventions differ in different parts of the world. Here is a Google query discussing the meaning of subtotal in ecommerce: [https://www.google.com/search?q=subtotal+in+ecommerce](https://www.google.com/search?q=subtotal+in+ecommerce)

**Comment by carlosa54** [comment]
@korycins thanks for the response, got it I missed stating the shipping price difference between the total and subtotal.

What I'm trying to bring up is that the current `subtotal` is returning a `TaxedMoney` that includes `tax, net and gross` and this includes the taxes.

Here's an [example order](https://test-org.saleor.cloud/graphql/#saleor/N4IgjiBc4K4KYCcCeACA8ggJogInALgIYCWANgM4AUAJMZpCgJI4CEAlCsADoB2KKAeyyJKdBrUwduffijq9ZKckXwxyC2eRgAjfAKKlOGxUQAeRmYv6EAtgJg98x2QF9n-HgQtXZt+4-cUN0tZAHMEAXJybx8UPwcnEP5gxRTZPQMYk0JzaVi4uwTAtMVPfCyreICkoMDwyOi82KrEnxLamWDgkAAaEAEoEAxsBDwiMnJekAA3QaauEHkoFAWAFQBmACkAawAtAA1GADYAWQAvbYAWE9WANRsAOQArAE18NBwTgCYH4gAGP4vGwAUUuABkAOrAgDuuwA4gBVfC7J6hM4vL63UgPRHXGz3Ba8FwgFxAA)

If an order has taxes but don't have shipping the subtotal is going to be the same as the total in the response (See below):
```json
"order": {
      "id": "T3JkZXI6Mzk4MTVmNjYtODM2Ni00YmE4LWEwZGUtZjgzY2VlNGU4MmVm",
      "status": "FULFILLED",
      "subtotal": {
        "tax": {
          "amount": 0.72
        },
        "net": {
          "amount": 10
        },
        "gross": {
          "amount": 10.72
        }
      },
      "total": {
        "tax": {
          "amount": 0.72
        },
        "net": {
          "amount": 10
        },
        "gross": {
          "amount": 10.72
        }
      }
    }
```

Since the `subtotal` should not include taxes, I think it should be a `Money` field with the value following the same definition that you posted

> By definition, the subtotal in ecommerce refers to the sum of all line prices multiplied by their respective quantities.

Let me know if the above explanation helps, thanks again for going over this and answering

**Comment by carlosa54** [comment]
@korycins although having the order subtotal as `TaxedMoney` instead of `Money` could cause confusion I understand that doing this change would introduce a breaking change to api users. Since they could already be using it correctly with `order.subtotal.net` I don't think that it's a huge problem leaving it as it is in the backend.

Given this I pushed a fix in https://github.com/saleor/saleor-dashboard/pull/3447 for the order subtotal display issue.

**Comment by korycins** [comment]
@carlosa54 
Why do you think `subtotal` should be a `Money` type? The subtotal is a sum of the lines multiplied by the quantity. Each line has its own taxes applied to it, so the subtotal is also a `taxedMoney`. See the query: [graphql-query](https://test-org.saleor.cloud/graphql/#saleor/N4IgjiBc4K4KYCcCeACA8ggJogInALgIYCWANgM4AUAJMZpCgJI4CEAlCsADoB2KKAeyyJKdBrUwduffijq9ZKckXwxyC2aWI845aYv74BRUgAUExAMZx9B-gHMEA8no13+hALYCYPfG4MAXwDFHXxbdxQvHz8Q2WCZRQSkgPIYACMjE044ogAPHMSDaN9-Iv5kgzDCyJLY8pRKxUdnchr3OrK7JsaArMJSdsV8oeLvUrie-mqIsZiuoLiWl1HFTsmAyoSEkAAaEAEoEAxsBDwiMnI9kAA3I9suEHkoFEeAFQBmACkAawAtAAajAAbABZABePwALKC3gA1TwAOQAVgBNfBoHCggBMiOIAAZ8ajPABRKEAGQA6iSAO5-ADiAFV8H9kfZwajsXDSIimTDPAjHrxAiBAkA).

The example that you shared looks okay. You have:

- `subtotal`: (net=10.00, gross=10.72)
- `shippingPrice`: (net=0, gross=0)
- `total`: `subtotal` + `shippingPrice`

The case when `total` is equal to `subtotal` only happens when the `shippingPrice` is 0.

Maybe I am wrong, but I think that you expect that dashboard will show subtotal, shippingPrice as a `net` prices and the separate line with `taxes` . Then `total` is `subtota-in-net` + `shipping-price-net` + `taxes`. Is that correct?
Something like(Based on above example):
```
- subtotal: 10 USD
- shippingPrice: 0 USD
- taxes: 0.72 USD
---
- total: 10.72 USD 
```

**Reply by carlosa54** [reply]
> Why do you think subtotal should be a Money type? The subtotal is a sum of the lines multiplied by the quantity. Each line has its own taxes applied to it, so the subtotal is also a taxedMoney. See the query: [graphql-query](https://test-org.saleor.cloud/graphql/#saleor/N4IgjiBc4K4KYCcCeACA8ggJogInALgIYCWANgM4AUAJMZpCgJI4CEAlCsADoB2KKAeyyJKdBrUwduffijq9ZKckXwxyC2aWI845aYv74BRUgAUExAMZx9B-gHMEA8no13+hALYCYPfG4MAXwDFHXxbdxQvHz8Q2WCZRQSkgPIYACMjE044ogAPHMSDaN9-Iv5kgzDCyJLY8pRKxUdnchr3OrK7JsaArMJSdsV8oeLvUrie-mqIsZiuoLiWl1HFTsmAyoSEkAAaEAEoEAxsBDwiMnI9kAA3I9suEHkoFEeAFQBmACkAawAtAAajAAbABZABePwALKC3gA1TwAOQAVgBNfBoHCggBMiOIAAZ8ajPABRKEAGQA6iSAO5-ADiAFV8H9kfZwajsXDSIimTDPAjHrxAiBAkA).

I think the main reason I thought that `subtotal` should be a `Money` type is because when displaying the `subtotal` value it will always be the `subtotal.net` when using `TaxedMoney` since the `Subtotal:` shouldn't include the taxes.


> Maybe I am wrong, but I think that you expect that dashboard will show subtotal, shippingPrice as a net prices and the separate line with taxes . Then total is subtota-in-net + shipping-price-net + taxes. Is that correct?

What I'm expecting in the dashboard is the `After` screenshot from the [PR I pushed](https://github.com/saleor/saleor-dashboard/pull/3447):
![image](https://user-images.githubusercontent.com/4250593/229547514-5fa0e08a-28ad-4bcb-9f71-f4de9700343b.png)

If the taxes are included in the `subtotal` when displaying the `subtotal` the only way for the Staff to know what was the subtotal without taxes is to calculate it

**Reply by korycins** [reply]
@carlosa54 
I understand your perspective. In most countries, it is common to show the subtotal as a `gross` value for shop prices. The `Taxes` field indicates the tax amount included in the total, so there is no need to know the net amount of the subtotal.

However, there is a missing setting that controls how values are displayed on the dashboard. The `Show gross prices in storefront` flag on Configuration→Taxes should have a different name because it is not only related to the storefront. Based on this flag, each order has a `displayGrossPrices` setting that determines whether the dashboard price section should operate on `net` or `gross` prices.

Based on the screenshot you provided, the `Before` view shows `displayGrossPrices` set to `True`. When `displayGrossPrices` is set to `False`, the displayed amounts should be in `net` value, and the `Taxes` section should not have the label `Included in prices` since taxes are shown as a separate section. This is the part that I believe is missing on the dashboard side as the dashboard always operates on `gross` amounts.

Would this solution solve your problem? Did I understand it correctly?

**Reply by carlosa54** [reply]
> In most countries, it is common to show the subtotal as a gross value for shop prices

@korycins I see, in that case the setting would make sense. I'm used to always seeing the `subtotal` as net in North America.

Thank you!

**Reply by korycins** [reply, marked as answer]
@carlosa54 Yes, that is correct. North America uses net as a subtotal, while Europe, for example, operates on gross.

During the refactoring of the dashboard, it seems that we lost this functionality. I'm pretty sure that previous versions of the dashboard (like 2.0+) were able to show amounts as net or gross based on these settings.

I've created a task to fix the issue:

- Core issue for incorrect field description: [https://github.com/saleor/saleor/issues/12492](https://github.com/saleor/saleor/issues/12492)
- Dashboard issue for incorrect field name: [https://github.com/saleor/saleor-dashboard/issues/3464](https://github.com/saleor/saleor-dashboard/issues/3464)
- Dashboard bug for not taking into account the `order.displayGrossPrices` setting: [https://github.com/saleor/saleor-dashboard/issues/3463](https://github.com/saleor/saleor-dashboard/issues/3463)

Thank you again for clarifying this and helping us find the issue.
